homeassistant/components/input_datetime/__init__.py

```diff
@@ -235,13 +235,15 @@ def __init__(self, config: dict) -> None:
 
         if self.has_date and self.has_time:
             current_datetime = dt_util.parse_datetime(initial)
-
+            assert current_datetime is not None  # has been validated by valid_initial
         elif self.has_date:
             date = dt_util.parse_date(initial)
+            assert date is not None  # has been validated by valid_initial
```

Comment:
Should we adjust `valid_inital` to return the parsed date instead? that would eliminate the need to assert.

Reply:
I made an attempt, adding a CONF_PARSED_INITIAL, but that broke all tests and I couldn't work out why.

We could also replace all the `assert` with `cast` instead.

Reply:
I rather leave it on the list instead of masking the issue until its fixed.

Reply:
I have moved the parsing out of valid_initial and into parse_initial_datetime.
This means the logic is now shared between the entity and the config validation.